File: mixin-base/src/main/java/net/yakclient/mixin/base/internal/loader/TargetClassLoader.java
package net.yakclient.mixin.base.internal.loader;

import net.yakclient.mixin.base.internal.bytecode.ByteCodeUtils;

import java.io.IOException;


public class TargetClassLoader extends ProxyClassLoader {
    private final PackageTarget target;

    public TargetClassLoader(ClassLoader parent, PackageTarget target) {
        super(parent);
        this.target = target;
    }

    @Override
    public Class<?> defineClass(String name, byte[] b) {
        return super.defineClass(name, b);
    }

    @Override
    protected final Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
        final var target = ClassTarget.of(name);
        if (!this.target.isTargetOf(target)) return super.loadClass(name, resolve);
        try {
            Class<?> c = this.findLoadedClass(name);

            if (c == null) {
                final var b = ByteCodeUtils.loadClassBytes(name);
                c = defineClass(name, b, 0, b.length);
            }

            if (resolve) this.resolveClass(c);

            return c;
        } catch (IOException e) {
            throw new ClassNotFoundException(e.getMessage());
        }
    }



    public PackageTarget getTarget() {
        return target;
    }
}

File: mixin-base/src/main/kotlin/net/yakclient/mixin/base/internal/loader/TargetClassLoader.kt
package net.yakclient.mixin.base.internal.loader

import net.yakclient.mixin.base.internal.bytecode.ByteCodeUtils.loadClassBytes
import net.yakclient.mixin.base.internal.loader.ClassTarget.Companion.of
import java.io.IOException

class TargetClassLoader(
    parent: ClassLoader,
    private val target: PackageTarget
) : ProxyClassLoader(parent) {

    @Throws(ClassNotFoundException::class)
    override fun loadClass(name: String, resolve: Boolean): Class<*> {
        val target = of(name)
        return if (!this.target.isTargetOf(target)) super.loadClass(name, resolve) else try {
            var c = findLoadedClass(name)
            if (c == null) {
                val b = loadClassBytes(name)
                c = defineClass(name, b, 0, b!!.size)
            }
            if (resolve) resolveClass(c)
            c
        } catch (e: IOException) {
            throw ClassNotFoundException(e.message)
        }
    }
}